pinot-server/src/main/java/org/apache/pinot/server/starter/helix/FreshnessBasedConsumptionStatusChecker.java

```diff
@@ -0,0 +1,102 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.server.starter.helix;
+
+import java.util.Set;
+import org.apache.pinot.core.data.manager.InstanceDataManager;
+import org.apache.pinot.core.data.manager.realtime.LLRealtimeSegmentDataManager;
+import org.apache.pinot.segment.local.data.manager.SegmentDataManager;
+import org.apache.pinot.spi.stream.LongMsgOffset;
+import org.apache.pinot.spi.stream.StreamPartitionMsgOffset;
+
+
+/**
+ * This class is used at startup time to have a more accurate estimate of the catchup period in which no query
+ * execution happens and consumers try to catch up to the latest messages available in streams.
+ * To achieve this, every time status check is called - {@link #getNumConsumingSegmentsNotReachedIngestionCriteria} -
+ * for each consuming segment, we check if either:
+ *   - the segment's latest ingested offset has reached the current stream offset that's
+ *   - the last ingested message is within {@link #_minFreshnessMs} of the current system time
+ */
+public class FreshnessBasedConsumptionStatusChecker extends IngestionBasedConsumptionStatusChecker {
+  private final long _minFreshnessMs;
+
+  public FreshnessBasedConsumptionStatusChecker(InstanceDataManager instanceDataManager, Set<String> consumingSegments,
+      long minFreshnessMs) {
+    super(instanceDataManager, consumingSegments);
+    _minFreshnessMs = minFreshnessMs;
+  }
+
+  private boolean isOffsetCaughtUp(StreamPartitionMsgOffset currentOffset, StreamPartitionMsgOffset latestOffset) {
+    if (currentOffset != null && latestOffset != null) {
+      if (currentOffset.compareTo(latestOffset) == 0) {
+        return true;
+      }
+      if (currentOffset instanceof LongMsgOffset && latestOffset instanceof LongMsgOffset) {
```

Comment:
Please change the interface to do this. We don't want any Kafka specific code in the pinot. It is not too hard to change the interface to add a comparator, with a default method.

Please, let us keep it clean.

Reply:
I can't think of how else to do this even in the interface. This issue only arises with Kafka because "latest offset" there is actually +1 of whatever your current offset is. The kinesis and pulsar implementations use a string and MessageId respectively to store the offset.

I can make a `getPreviousOffset` method in the `LLRealtimeSegmentDataManager`, but then I just move this logic there. 

I can put a `fetchPreviousOffset` in `StreamMetadataProvider`, but then I'll just have to raise a `NotImplementedException` by default and later go update our own custom `StreamMetadataProvider`.

Reply:
Here is one option: add a method called `isNextOffset(streamMsgOffset)` you can also add a method (if need be) like `isNextOffsetSupported()` or some such, so that you can make sure you call it only if it is supported. the default can throw exceptions

Reply:
@sajjad-moradi Seems we will have the same problem in the offset based status checker when the ingestion rate is very low? If we never consume anything since the server restarts, it will never reach the latest offset since it is exclusive.

I feel we run into this issue because of the `fetchStreamPartitionOffset()` implementation for kafka 2.0. For all other implementations, it will return the last available offset, but for kafka 2.0 it returns the next offset after the last available offset. We should fix the contract for this method.

IMO we should have a way to get the last available offset, and return `null` if there is no message available in the stream. That's the only way to ensure the consumption already catches up

Reply:
This feels like a bigger issue than I want to tackle in this PR. I've removed the kafka specific code for now. We have our own stream ingestion plugin where I will update the `fetchStreamPartitionOffset` to return `- 1`.

Reply:
> @sajjad-moradi Seems we will have the same problem in the offset based status checker when the ingestion rate is very low? If we never consume anything since the server restarts, it will never reach the latest offset since it is exclusive.
> 
> I feel we run into this issue because of the `fetchStreamPartitionOffset()` implementation for kafka 2.0. For all other implementations, it will return the last available offset, but for kafka 2.0 it returns the next offset after the last available offset. We should fix the contract for this method.
> 
> IMO we should have a way to get the last available offset, and return `null` if there is no message available in the stream. That's the only way to ensure the consumption already catches up

That's correct. I created this issue for it: https://github.com/apache/pinot/issues/9291